### PR TITLE
kubelet: migrate pkg/kubelet/certificate to structured logging

### PR DESCRIPTION
**Pull request by SataQiu**
#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

Ref #98976

#### Does this PR introduce a user-facing change?

```release-note
Migrate `pkg/kubelet/certificate` to structured logging
ACTION REQUIRED: Exit code changed from 255 to 1
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/certificate/transport.go [hunk 1]
```diff
@@ -105,18 +106,20 @@ func addCertRotation(stopCh <-chan struct{}, period time.Duration, clientConfig
 				// the certificate has been deleted from disk or is otherwise corrupt
 				if now.After(lastCertAvailable.Add(exitAfter)) {
 					if clientCertificateManager.ServerHealthy() {
-						klog.Fatalf("It has been %s since a valid client cert was found and the server is responsive, exiting.", exitAfter)
+						klog.ErrorS(nil, "It has been sometime since a valid client cert was found and the server is responsive, exiting.", "exitAfter", exitAfter)
```
Comment (endocrimes) [inline pkg/kubelet/certificate/transport.go:109]: We could maybe improve this message a little, maybe with something like:

```
klog.ErrorS(nil, "Shutting down as no valid client cert was found and the server was responsive", "lastCertificateAvailabilityTime", lastCertAvailable, "shutdownThreshold", exitAfter)
```

**Comment by ehashman** [comment]
Because this is updating the exit code of a Fatal log call, we need an `ACTION REQUIRED:` release note.

/hold

File: pkg/kubelet/certificate/transport.go [hunk 2]
```diff
@@ -129,7 +132,7 @@ func addCertRotation(stopCh <-chan struct{}, period time.Duration, clientConfig
 		}
 		lastCert = curr
 
-		klog.Infof("certificate rotation detected, shutting down client connections to start using new credentials")
+		klog.InfoS("certificate rotation detected, shutting down client connections to start using new credentials")
```
Comment (adisky) [inline pkg/kubelet/certificate/transport.go:135]:
```suggestion
		klog.InfoS("Certificate rotation detected, shutting down client connections to start using new credentials")
```

File: pkg/kubelet/certificate/transport.go (hunk 1 above)
Comment (SataQiu) [reply pkg/kubelet/certificate/transport.go]: Thanks, the code has been updated @endocrimes

**Comment by SataQiu** [comment]
> Because this is updating the exit code of a Fatal log call, we need an `ACTION REQUIRED:` release note.
> 
> /hold

The `ACTION REQUIRED` was added to the release note.
And all the suggestions were applied.

**Comment by SataQiu** [comment]
@ehashman  @serathius Can we proceed with this PR?

**Comment by SataQiu** [comment]
Follow the robot's instruction
/assign @liggitt

**Review by ehashman — COMMENTED** [review]
/hold cancel
/lgtm

**Comment by SataQiu** [comment]
kindly ping @liggitt 
@ehashman had already left `LGTM`.

File: pkg/kubelet/certificate/transport.go [hunk 3]
```diff
@@ -105,18 +106,20 @@ func addCertRotation(stopCh <-chan struct{}, period time.Duration, clientConfig
 				// the certificate has been deleted from disk or is otherwise corrupt
 				if now.After(lastCertAvailable.Add(exitAfter)) {
 					if clientCertificateManager.ServerHealthy() {
-						klog.Fatalf("It has been %s since a valid client cert was found and the server is responsive, exiting.", exitAfter)
+						klog.ErrorS(nil, "No valid client certificate is found and the server is responsive, exiting.", "lastCertificateAvailabilityTime", lastCertAvailable, "shutdownThreshold", exitAfter)
+						os.Exit(1)
 					} else {
-						klog.Errorf("It has been %s since a valid client cert was found, but the server is not responsive. A restart may be necessary to retrieve new initial credentials.", exitAfter)
+						klog.ErrorS(nil, "No valid client certificate is found but the server is not responsive. A restart may be necessary to retrieve new initial credentials.", "lastCertificateAvailabilityTime", lastCertAvailable, "shutdownThreshold", exitAfter)
```
Comment (liggitt) [inline pkg/kubelet/certificate/transport.go:112]: are the keys selected for messages like these considered a stable API log consumers can write extraction/monitoring around?

File: pkg/kubelet/certificate/transport.go [hunk 4]
```diff
@@ -105,18 +106,20 @@ func addCertRotation(stopCh <-chan struct{}, period time.Duration, clientConfig
...
 			} else {
 				// the certificate is expired
 				if now.After(curr.Leaf.NotAfter) {
 					if clientCertificateManager.ServerHealthy() {
-						klog.Fatalf("The currently active client certificate has expired and the server is responsive, exiting.")
+						klog.ErrorS(nil, "The currently active client certificate has expired and the server is responsive, exiting.")
+						os.Exit(1)
 					} else {
-						klog.Errorf("The currently active client certificate has expired, but the server is not responsive. A restart may be necessary to retrieve new initial credentials.")
+						klog.ErrorS(nil, "The currently active client certificate has expired, but the server is not responsive. A restart may be necessary to retrieve new initial credentials.")
```
Comment (liggitt) [inline pkg/kubelet/certificate/transport.go:122]: passing a nil error to ErrorS seems strange... should this do something like `klog.ErrorS(errors.New("The currently active client..."))` instead?

File: pkg/kubelet/certificate/transport.go (hunk 3 above)
Comment (ehashman) [reply pkg/kubelet/certificate/transport.go:112]: Not as of beta but we're moving in that direction for GA. Right now we're trying to figure out a good (loose, non-enforced) schema that we can later enforce.

File: pkg/kubelet/certificate/transport.go (hunk 4 above)
Comment (ehashman) [reply pkg/kubelet/certificate/transport.go:122]: Docs on calling `ErrorS` with `nil`: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#using-errors

> Calling ErrorS with nil as error is semi-acceptable if there is error condition that deserves a stack trace at this origin point. For expected errors (errors that can happen during routine operations) please consider using klog.InfoS and pass error in err key instead.

I think this is an unexpected error so this is ok?

Comment (liggitt) [reply pkg/kubelet/certificate/transport.go:122]: it looks like the klog implementation gracefully omits the err=nil from the output:

https://github.com/kubernetes/kubernetes/blob/c8fe1d9c0699f8d1fd93f47f2a62a4c83ccf5b6f/vendor/k8s.io/klog/v2/klog.go#L803-L806

but the structured logger always includes an err param, even if nil:

https://github.com/kubernetes/kubernetes/blob/c8fe1d9c0699f8d1fd93f47f2a62a4c83ccf5b6f/staging/src/k8s.io/component-base/logs/json/json.go#L171-L173

is that expected by structured log consumers?

Comment (ehashman) [reply pkg/kubelet/certificate/transport.go:122]: Yes, that's right. We just get the log message with no named error, so it's equivalent to the current behaviour.

**Pull request merged**